src/NServiceBus.Gateway/GatewayExtensions.cs

```diff
@@ -29,5 +29,19 @@ public static bool IsLegacyGatewayMessage(this IDictionary<string, string> heade
 
             return legacyMode;
         }
+
+        /// <summary>
+        ///     encode headers for reverse proxy
+        /// </summary>
+        public static IDictionary<string, string> EncodeHeadersForReverseProxy(this IDictionary<string, string> headers)
+        {
+            var encodedHeaders = new Dictionary<string, string>();
+            foreach (var header in headers)
+            {
+                encodedHeaders.Add(header.Key.Replace(".","-"),header.Value);
```

Comment:
The encoder is different in behavior then the decoder.

Reply:
Only headers starting with `NServiceBus.` should then be addressed. Also, this would mean that we *must* make sure that we always use dots in our header keys.

Reply:
@ramonsmits It's not that we have to use `.`, it's that we can't use `-` in anywhere in the key, or it will incorrectly be turned into `.` on the decoding side.

Given that we don't have any sort of enforcement around this, I'm not sure how tenable the approach being taken here is.

Reply:
I think it should only be updating headers that start `NServiceBus.`. I did a bit more playing around to get different headers, and it looks like my current approach won't work when using a databus; that adds headers with `-` in them and also `_`, which is another problem character for nginx.

I think that something like this could work:

If the header begins `NServiceBus.`, then replace with `NServiceBus-{base64 encoded header name}`. As long as all NServiceBus headers start with `NServiceBus.` then that should be ok. Any thoughts on that approach?